torchqdynamics/solver.py

```diff
@@ -104,11 +104,20 @@ def save(self, y: Tensor):
         self._save_exp_ops(y)
         self.save_counter += 1
 
+    def save_initial(self, y: Tensor):
+        if self.options.save_states and self.next_tsave() == 0.0:
+            self.save(y)
+
     def save_final(self, y: Tensor):
-        if self.options.save_states:
-            self._save_y(y)
-        else:
+        if not self.options.save_states:
             self.y_save = y
+        elif self.save_counter < len(self.t_save):
+            self._save_y(y)
+
+        if self.save_counter < len(self.t_save):
+            self._save_exp_ops(y)
+
+        self.save_counter += 1
```

Comment:
Again, this is too specific to the adaptive ODE solver, it should not be in the base class.

I think we should actually remove `save_final()` and modify the base `Solver` class `_save_y()` code to automatically handle the final state saving. Currently, the `Propagator` code is bugged for example, and with the current design we need to call `self.save_final(y)` after every solver, which is annoying. Here's the fix:
```python
def _save_y(self, y: Tensor):
    if self.options.save_states:
        self.y_save[..., self.save_counter, :, :] = y
    # otherwise save only final state
    elif self.save_counter == len(self.t_save):
        self.y_save = y
```

Then we just need to
- call `self.save()` once again in the fixed step forward code (for the last time point):
   ```diff
   # save final time step (`t` goes `0.0` to `t_save[-1]` excluded)
   - self.save_final(y)
   + self.save(y)
   ```
- fix the `next_tsave` to return `None` if there is no more time steps:
   ```diff
   def next_tsave(self) -> float | None:
   -     return self.t_save[self.save_counter]
   +     return self.t_save[self.save_counter] if self.save_counter < len(self.t_save) else None
   ```
- at the end of the adaptive step forward code (for the last time point):
   ```diff
   # save last state if not already done
   - self.save_final(y)
   + if self.next_t_save() is not None:
   +    self.save(y)
   ```

Wdyt?